// @sap/ux-specification/dist/src/sync/v2/export/controls/DataField.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const decoration_1 = require("../../../common/decoration");
const common_1 = require("../../../common");
const getSmartFieldControlId = (baseId, idBreadcrumbs, breadcrumbs) => {
    const breadCrumbId = breadcrumbs.filter((crumb) => !idBreadcrumbs.includes(crumb));
    const titleTrail = breadCrumbId[0].replace('@com.sap', 'com.sap');
    const dataFieldCrumb = breadcrumbs.slice(-1)[0];
    const formId = dataFieldCrumb.includes('::') ? dataFieldCrumb.split('DataField::')[1] : dataFieldCrumb;
    const breadcrumbsTrail = [titleTrail, formId, 'GroupElement'];
    return decoration_1.buildControlIdFromParent(baseId, breadcrumbsTrail);
};
class DataField {
}
__decorate([
    decoration_1.syncRule({
        flex: {
            controlId: getSmartFieldControlId,
            controlType: () => common_1.ControlType.SmartGroup,
            createFlexChange: decoration_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], DataField.prototype, "visible", void 0);
exports.DataField = DataField;
