// @sap/ux-specification/dist/src/sync/v4/import/fragment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const v4_1 = require("../../../v4");
const path_1 = require("path");
/**
 * Determines a related fragment file and returns its content
 */
function readFragmentFile(fragmentName, fragmentFiles, manifestPart, appId) {
    let fragmentData;
    if (fragmentName) {
        const fragmentId = fragmentName.substr(appId.length + 1);
        const pathParts = fragmentId.split('.');
        pathParts[pathParts.length - 1] = pathParts[pathParts.length - 1] + '.fragment.xml';
        const extensionPath = path_1.join('webapp', ...pathParts);
        for (let index = 0; index < fragmentFiles.length; index++) {
            const file = fragmentFiles[index];
            if (file.dataSourceUri.endsWith(extensionPath)) {
                fragmentData = {
                    manifestPart,
                    fragmentFileContent: file.fileContent
                };
                break;
            }
        }
    }
    return fragmentData;
}
/**
 * Reads an extension from a given segment of a V4 target's controlConfiguration
 */
function readControlConfigurationExtension(manifest, target, targetSegment, appId, fragmentFiles) {
    const fragmentData = [];
    const targets = manifest[v4_1.ManifestSection.ui5].routing.targets;
    for (const pageKey in targets) {
        if (pageKey === target) {
            const targetSettings = targets[pageKey].options && targets[pageKey].options.settings;
            if (targetSettings) {
                const controlConfiguration = targetSettings.controlConfiguration;
                if (controlConfiguration) {
                    for (const configKey in controlConfiguration) {
                        const segments = controlConfiguration[configKey][targetSegment];
                        if (segments) {
                            for (const key in segments) {
                                if (segments[key]['template']) {
                                    //extension as part of controlConfiguration
                                    const fragment = readFragmentFile(segments[key]['template'], fragmentFiles, segments[key], appId);
                                    if (fragment) {
                                        fragmentData.push(fragment);
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    return fragmentData;
}
/**
 * Note: header or footer actions are not linked to a fragment, thus cannot be found here
 */
/**
 * Reads the header facet extensions
 */
function readHeaderFacetExtensions(manifest, target, appId, fragmentFiles) {
    var _a, _b, _c, _d;
    const fragmentData = [];
    const targets = manifest[v4_1.ManifestSection.ui5].routing.targets;
    for (const pageKey in targets) {
        if (pageKey === target) {
            const targetSettings = targets[pageKey].options && targets[pageKey].options.settings;
            if (targetSettings) {
                // Read extensions for header facets
                const configBasedFacets = (_b = (_a = targetSettings.controlConfiguration) === null || _a === void 0 ? void 0 : _a['@com.sap.vocabularies.UI.v1.HeaderFacets']) === null || _b === void 0 ? void 0 : _b.facets;
                const contentBasedFacets = (_d = (_c = targetSettings.content) === null || _c === void 0 ? void 0 : _c.header) === null || _d === void 0 ? void 0 : _d.facets;
                const headerFacets = Object.assign({}, configBasedFacets, contentBasedFacets);
                if (headerFacets) {
                    for (const key in headerFacets) {
                        const fragment = readFragmentFile(headerFacets[key]['name'], fragmentFiles, headerFacets[key], appId);
                        if (fragment) {
                            fragmentData.push(fragment);
                        }
                    }
                }
            }
        }
    }
    return fragmentData;
}
exports.readHeaderFacetExtensions = readHeaderFacetExtensions;
/**
 * Reads the section extensions (merge content with control configuration, content-based definitions win)
 */
function readSectionExtensions(manifest, target, appId, fragmentFiles) {
    var _a, _b, _c, _d;
    const fragmentData = [];
    const targets = manifest[v4_1.ManifestSection.ui5].routing.targets;
    for (const pageKey in targets) {
        if (pageKey === target) {
            const targetSettings = targets[pageKey].options && targets[pageKey].options.settings;
            if (targetSettings) {
                // Read extensions for sections
                const contentBasedSections = (_b = (_a = targetSettings.content) === null || _a === void 0 ? void 0 : _a.body) === null || _b === void 0 ? void 0 : _b.sections;
                const configBasedSections = (_d = (_c = targetSettings.controlConfiguration) === null || _c === void 0 ? void 0 : _c['@com.sap.vocabularies.UI.v1.Facets']) === null || _d === void 0 ? void 0 : _d.sections;
                const sections = Object.assign({}, configBasedSections, contentBasedSections);
                if (sections) {
                    for (const key in sections) {
                        const fragment = readFragmentFile(sections[key]['name'], fragmentFiles, sections[key], appId);
                        if (fragment) {
                            fragmentData.push(fragment);
                        }
                    }
                }
            }
        }
    }
    return fragmentData;
}
exports.readSectionExtensions = readSectionExtensions;
/**
 * Reads the columns extensions (from control configuration)
 */
function readColumnExtensions(manifest, target, appId, fragmentFiles) {
    return readControlConfigurationExtension(manifest, target, 'columns', appId, fragmentFiles);
}
exports.readColumnExtensions = readColumnExtensions;
/**
 * Reads the filter fields extensions (from control configuration)
 */
function readFilterFieldsExtensions(manifest, target, appId, fragmentFiles) {
    return readControlConfigurationExtension(manifest, target, 'filterFields', appId, fragmentFiles);
}
exports.readFilterFieldsExtensions = readFilterFieldsExtensions;
